prov/sm2/src/sm2_progress.c

```diff
@@ -110,8 +166,35 @@ static int sm2_start_common(struct sm2_ep *ep,
 			"Unable to process rx completion\n");
 	}
 
-	if (!(xfer_entry->hdr.proto_flags & SM2_UNEXP))
+	if (!sm2_proto_imm_send_comp(xfer_entry->hdr.proto))
+		xfer_entry->hdr.proto_flags |= SM2_GENERATE_COMPLETION;
+
+	if (!(xfer_entry->hdr.proto_flags & SM2_UNEXP)) {
 		sm2_fifo_write_back(ep, xfer_entry);
+	} else if (!sm2_proto_imm_send_comp(xfer_entry->hdr.proto)) {
+		/* The receiver has completed an unexpected receive with
+		 * an xfer_entry from the xfer_ctx_pool but the sender
+		 * has not generated a send completion yet. So we need to
+		 * send a new xfer_entry from the receiver's freestack
+		 * with the SM2_GENERATE_COMPLETION flag, so that the sender can
+		 * generate a send completion */
+		ret = sm2_pop_xfer_entry(ep, &new_xfer_entry);
+		if (ret) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Unable to send xfer_entry back to "
+				"sender. Sender has requested "
+				"FI_DELIVERY_COMPLETE but will not be "
+				"able to generate a send "
+				"completion!\n");
+			return ret;
```

Comment:
These seems like a dangerous way to handle this issue... just error out the sender because the app is sending too many things of its own? If you want to keep that fix for the later patch that fixes the unexpected messaging, that's ok, but maybe include that in your comment regarding the unexpected messaging handling future plans.

Reply:
Makes sense. Add a TODO to add a retry.